keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,938 @@
+
+6. We will also add a poststarthook for the apiserver to ensure that it does not start serving requests until we are done creating/updating SV objects
+
+[handler chain]:https://github.com/kubernetes/kubernetes/blob/fc8f5a64106c30c50ee2bbcd1d35e6cd05f63b00/staging/src/k8s.io/apiserver/pkg/server/config.go#L639
+
+#### StorageVersion enhancement needed
+
+StorageVersion API currently tells us whether a particular StorageVersion can be read from etcd by the listed apiserver. We will enhance this API to also include apiserver ID of the server that can serve this StorageVersion.
+
+With the enhancement, the new [ServerStorageVersion](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/apis/apiserverinternal/types.go#L62-L73) object will have this structure
+
+```
+type ServerStorageVersion struct {
+  // The ID of the reporting API server.
+  APIServerID string
+	
+  // The API server encodes the object to this version 
+  // when persisting it in the backend (e.g., etcd).
+  EncodingVersion string
+
+  // The API server can decode objects encoded in these versions.
+  // The encodingVersion must be included in the decodableVersions.
+  DecodableVersions []string
+
+  // Versions that can be served by the reporting API server
+  ServedVersions []string
+}
+```
+
+#### Identifying destination apiserver's network location
+
+We will be performing dual writes of the ip and port information of the apiservers in:
+
+1. A clone of the [endpoint reconciler's masterlease](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/reconcilers/lease.go) which would be read by apiservers to proxy the request to a peer. We will use a separate reconciler loop to do these writes to avoid modifying the existing endpoint reconciler
+
+2. [APIServerIdentity Lease object](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/instance.go#L559-L577) for users to view this information for debugging
+
+3. We will use an egress dialer for network connections made to peer kube-apiservers. For this, will create a new type for the network context to be used for peer kube-apiserver connections ([xref](https://github.com/kubernetes/kubernetes/blob/release-1.27/staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go#L56-L71))
+
+#### Proxy transport between apiservers and authn
+
+For the mTLS between source and destination apiservers, we will do the following
+
+1. For server authentication by the client (source apiserver) : the client needs to validate the [server certs](https://github.com/kubernetes/kubernetes/blob/release-1.27/staging/src/k8s.io/apiserver/pkg/server/options/serving.go#L59) (presented by the destination apiserver), for which it will 
+   1. look at the CA bundle of the authority that signed those certs. We will introduce a new flag --peer-ca-file that **is required be passed (for this feature)** to the kube-apiserver that will be used to verify the presented server certs
```

Comment:
Re `--peer-ca-file that **is required be passed (for this feature)**`.  I think we should allow the feature to be enabled even if there is no networking enabled between apiservers to satisfy https://github.com/kubernetes/enhancements/pull/4015#issuecomment-1583094402.  In such a configuration the difference between having the feature on and off is that when it is on 503s instead of 404s.

Reply:
Addresses in https://github.com/kubernetes/enhancements/pull/4074